lib/Dialect/ESI/runtime/cpp/include/esi/cosim/backends/capnp.h

```diff
@@ -0,0 +1,301 @@
+
+  auto structValue = value.as<DynamicStruct>();
+  for_each(refl::reflect(ret).members, [&](auto member) {
+    auto fieldValue = structValue.get(member.name);
+    using ValueType = typename decltype(member)::value_type;
+    auto castValue = fieldValue.template as<ValueType>();
+    member(ret) = castValue;
+  });
+
+  return ret;
+}
+
+// A function like above, but which uses Capnproto dynamic message building
+// together with the refl-cpp reflection to constuct the Capnp message.
+template <typename TESIType>
+void toCapnp(TESIType &value, DynamicValue::Builder &message) {
+  if (message.getType() != DynamicValue::STRUCT)
+    throw std::runtime_error(
+        "Expected a struct value (all ESI types should be capnp structs)");
+
+  auto structMessage = message.as<DynamicStruct>();
+  for_each(refl::reflect(value).members, [&](auto member) {
+    auto &fieldValue = member(value);
+    structMessage.set(member.name, fieldValue);
+  });
+}
+
+namespace detail {
+// Custom type to hold the interface descriptions because i can't for the life
+// of me figure out how to cleanly keep capnproto messages around...
+struct EsiDpiInterfaceDesc {
+  std::string endpointID;
+  uint64_t sendTypeID;
+  uint64_t recvTypeID;
+};
+} // namespace detail
+
+using EsiDpiInterfaceDesc = detail::EsiDpiInterfaceDesc;
+
+template <typename WriteType, typename ReadType>
+class CapnpReadWritePort;
+
+template <typename WriteType>
+class CapnpWritePort;
+
+template <typename ReadType>
+class CapnpReadPort;
+
+class CapnpBackend {
+public:
+  // Using directives to point the base class implementations to the cosim
+  // port implementations.
+
+  template <typename WriteType, typename ReadType>
+  using ReadWritePort = CapnpReadWritePort<WriteType, ReadType>;
+
+  template <typename WriteType>
+  using WritePort = CapnpWritePort<WriteType>;
+
+  template <typename ReadType>
+  using ReadPort = CapnpReadPort<ReadType>;
+
+  CapnpBackend(const std::string &host, uint64_t hostPort) {
+    ezClient = std::make_unique<capnp::EzRpcClient>(host, hostPort);
+    dpiClient = std::make_unique<CosimDpiServer::Client>(
+        ezClient->getMain<CosimDpiServer>());
+
+    list();
+  }
+
+  // Returns a list of all available endpoints.
+  const std::vector<detail::EsiDpiInterfaceDesc> &list() {
+    if (endpoints.has_value())
+      return *endpoints;
+
+    // Query the DPI server for a list of available endpoints.
+    auto listReq = dpiClient->listRequest();
+    auto ifaces = listReq.send().wait(ezClient->getWaitScope()).getIfaces();
+    endpoints = std::vector<detail::EsiDpiInterfaceDesc>();
+    for (auto iface : ifaces) {
+      detail::EsiDpiInterfaceDesc desc;
+      desc.endpointID = iface.getEndpointID().cStr();
+      desc.sendTypeID = iface.getSendTypeID();
+      desc.recvTypeID = iface.getRecvTypeID();
+      endpoints->push_back(desc);
+    }
+
+    // print out the endpoints
+    for (auto ep : *endpoints) {
```

Comment:
Debug code?

Reply:
Added this to match the printout of endpoints that the ESI `CosimBase` implementation does. I think i'd actually prefer to keep this in for now, as a sanity and early warning system. Can always be removed down the road.